shared/src/main/scala/io/udash/i18n/TranslationProvider.scala

```
@@ -0,0 +1,22 @@
+package io.udash.i18n
+
+import scala.concurrent.Future
+import scala.util.Try
+
+trait TranslationProvider {
+  private val argRegex = """\{(\d*)\}""".r
+
+  /** Basing on provided translation key, arguments and language it should created `Future` containing translated text. */
+  def translate(key: String, argv: Any*)(implicit lang: Lang): Future[Translated]
+
+  protected def putArgs(template: String, argv: Any*): Translated = {
+    val args = argv.map(_.toString).lift
+    var prevN = -1
+    val result = argRegex.replaceSomeIn(template, m => {
+      val n = Try(m.group(1).toInt).getOrElse(prevN + 1)
```

I'd throw a runtime exception, developer will notice that easily, otherwise we should check the bundles for invalid keys on app start-up.